### PR TITLE
Added mcap logging functionality

### PR DESCRIPTION
**Pull request by yaswanth1701**
Tested on: Ubuntu 22 and gazebo 11.
Solves #24, Replaced logging using junit format to mcap. Now instead of calculating performance metrics in the source file and recording the maximum error only, model states are logged to reduce simulator dependency. A custom message `boxes_msg.proto` is defined to store all the necessary data using mcap API which can used in the future for other simulators too.

To test updates:
```bash

cd benchmark
git submodule update --init --recursive
mkdir build
cd build
cmake ..
make
make test
```
Individual test:
```bash
./BENCHMARK_boxes_dt
```
NOTE: Currently, states of a single model are logged in both tests `BENCHMARK_boxes_dt` & `BENCHMARK_boxes_model_count`, however `boxes_msg.proto` is defined in such a way that it can log all model states in a single message if needed.

### CODE REVIEW

File: proto/boxes_msg.proto [hunk 1]
```diff
@@ -0,0 +1,62 @@
+syntax = "proto3";
+
+package benchmark_proto;
+message  Pose{
+    // position in world frame
+    Vector3 position = 1;
+
+    // orientation of boyd wrt world frame
+    Vector3 orientation = 2;
```
Comment (scpeters) [inline proto/boxes_msg.proto:9]: I think a quaternion would be better here, similar to [gz::msgs::Pose](https://github.com/gazebosim/gz-msgs/blob/gz-msgs10/proto/gz/msgs/pose.proto#L39) (see also https://github.com/gazebosim/gz-msgs/blob/gz-msgs10/proto/gz/msgs/quaternion.proto)

and you should be able to use gz-math7 python bindings in the postprocessing script for the quaternion math

**Comment by scpeters** [comment]
I'm seeing a linking error with abseil when building this on my macOS machine:

~~~
[ 31%] Linking CXX executable BENCHMARK_boxes_dt
ld: warning: ignoring duplicate libraries: '-ldl', '-lm', 'libgtest.a'
ld: Undefined symbols:
  void absl::lts_20240116::log_internal::LogMessage::CopyToEncodedBuffer<(absl::lts_20240116::log_internal::LogMessage::StringType)0>(std::__1::basic_string_view<char, std::__1::char_traits<char>>), referenced from:
      absl::lts_20240116::log_internal::LogMessage& absl::lts_20240116::log_internal::LogMessage::operator<<<19>(char const (&) [19]) in boxes.cc.o
      absl::lts_20240116::log_internal::LogMessage& absl::lts_20240116::log_internal::LogMessage::operator<<<48>(char const (&) [48]) in boxes.cc.o
      absl::lts_20240116::log_internal::LogMessage& absl::lts_20240116::log_internal::LogMessage::operator<<<59>(char const (&) [59]) in boxes.cc.o
  absl::lts_20240116::log_internal::LogMessage& absl::lts_20240116::log_internal::LogMessage::operator<<<unsigned long, 0>(unsigned long const&), referenced from:
      absl::lts_20240116::log_internal::LogMessage::operator<<(unsigned long) in boxes.cc.o
  absl::lts_20240116::log_internal::LogMessageFatal::LogMessageFatal(char const*, int, std::__1::basic_string_view<char, std::__1::char_traits<char>>), referenced from:
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      char* google::protobuf::Arena::CreateArray<char>(google::protobuf::Arena*, unsigned long) in boxes.cc.o
      google::protobuf::RepeatedField<double>::elements() const in boxes.cc.o
      ...
  absl::lts_20240116::log_internal::LogMessageFatal::~LogMessageFatal(), referenced from:
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::internal::ArenaAlignAs(unsigned long) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      google::protobuf::RepeatedField<double>::GrowNoAnnotate(int, int) in boxes.cc.o
      ...
  std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<void const*, void const*>(void const*, void const*, char const*), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Pose const*, benchmark_proto::Pose*>(benchmark_proto::Pose const* const&, benchmark_proto::Pose* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Twist const*, benchmark_proto::Twist*>(benchmark_proto::Twist const* const&, benchmark_proto::Twist* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Vector3 const*, benchmark_proto::Vector3*>(benchmark_proto::Vector3 const* const&, benchmark_proto::Vector3* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Quaternion const*, benchmark_proto::Quaternion*>(benchmark_proto::Quaternion const* const&, benchmark_proto::Quaternion* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Boxes_msg_Box_msg const*, benchmark_proto::Boxes_msg_Box_msg*>(benchmark_proto::Boxes_msg_Box_msg const* const&, benchmark_proto::Boxes_msg_Box_msg* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_NEImpl<benchmark_proto::Boxes_msg const*, benchmark_proto::Boxes_msg*>(benchmark_proto::Boxes_msg const* const&, benchmark_proto::Boxes_msg* const&, char const*) in boxes_msg.pb.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_EQImpl<google::protobuf::Arena*, google::protobuf::Arena*>(google::protobuf::Arena* const&, google::protobuf::Arena* const&, char const*) in boxes_msg.pb.cc.o
      ...
  std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<long long, long long>(long long, long long, char const*), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_GTImpl<int, int>(int const&, int const&, char const*) in boxes.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_GEImpl<int, int>(int const&, int const&, char const*) in boxes.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_LTImpl<int, int>(int const&, int const&, char const*) in boxes.cc.o
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_EQImpl<int, int>(int const&, int const&, char const*) in boxes.cc.o
  std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<unsigned long long, unsigned long long>(unsigned long long, unsigned long long, char const*), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::Check_LEImpl<unsigned long, unsigned long>(unsigned long const&, unsigned long const&, char const*) in boxes.cc.o
  absl::lts_20240116::log_internal::CheckOpMessageBuilder::ForVar2(), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<unsigned long long, long long>(unsigned long long, long long, char const*) in boxes.cc.o
  absl::lts_20240116::log_internal::CheckOpMessageBuilder::NewString(), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<unsigned long long, long long>(unsigned long long, long long, char const*) in boxes.cc.o
  absl::lts_20240116::log_internal::CheckOpMessageBuilder::CheckOpMessageBuilder(char const*), referenced from:
      std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::lts_20240116::log_internal::MakeCheckOpString<unsigned long long, long long>(unsigned long long, long long, char const*) in boxes.cc.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [BENCHMARK_boxes_dt] Error 1
make[1]: *** [CMakeFiles/BENCHMARK_boxes_dt.dir/all] Error 2
make: *** [all] Error 2
~~~

~~~
==> abseil: stable 20240116.2 (bottled), HEAD
~~~

**Comment by yaswanth1701** [comment]
Sorry  @scpeters but I don't have a macOS machine available with me currently to test the changes. I found a similar issue here https://github.com/protocolbuffers/protobuf/issues/12292  and made some changes by referring to the discussion of the issue to cmakelist if you don't mind could you please try building again?

File: CMakeLists.txt [hunk 2]
```diff
@@ -3,8 +3,21 @@ project(benchmark_gazebo)
 enable_testing()
 
 include(FindBoost)
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(zstd REQUIRED IMPORTED_TARGET libzstd)
+pkg_check_modules(lz4 REQUIRED IMPORTED_TARGET liblz4)
+find_package(Protobuf CONFIG)
```
Comment (scpeters) [inline CMakeLists.txt:9]:
```suggestion
find_package(protobuf CONFIG)
```
this works if I change the capitalization

Comment (yaswanth1701) [reply CMakeLists.txt:9]: oh, that's nice, it works on Ubuntu 22 too, I think the issue is with the version of protobuf, the official [documentation](https://cmake.org/cmake/help/latest/module/FindProtobuf.html) uses the capitalized version in CMake. Can I know which protobuf version you are using?

Comment (scpeters) [reply CMakeLists.txt]: I'm using protobuf from homebrew-core on macOS, which is currently version 26.1 on my machine, and will be 27.0 as soon as I upgrade

**Comment by scpeters** [comment]
on my macOS machine I get python errors when running the `csv_*` tests with `ctest -VV`:

~~~
4:   File "/Users/scpeters/clone/benchmark/tools/mcap_to_csv.py", line 3, in <module>
4:     from mcap_protobuf.decoder import DecoderFactory
4: ModuleNotFoundError: No module named 'mcap_protobuf'
~~~

I'm guessing there's a python path issue, but I haven't tried to sort it out yet

**Comment by yaswanth1701** [comment]
Pushed some changes, it should resolve the path issues

**Review by scpeters — COMMENTED** [review]
unfortunately the `css_*` tests are still failing for me, being unable to find the `lz4` python module

~~~
test 2
    Start 2: csv_BENCHMARK_boxes_dt

2: Test command: /usr/local/bin/python3 "/Users/scpeters/ws/benchmark/src/benchmark/tools/mcap_to_csv.py" "BENCHMARK_boxes_dt"
2: Working Directory: /Users/scpeters/ws/benchmark/src/benchmark/build
2: Test timeout computed to be: 10000000
2: Traceback (most recent call last):
2:   File "/Users/scpeters/ws/benchmark/src/benchmark/tools/mcap_to_csv.py", line 6, in <module>
2:     from mcap.reader import make_reader
2:   File "/Users/scpeters/ws/benchmark/src/benchmark/mcap/python/mcap/mcap/reader.py", line 37, in <module>
2:     from .stream_reader import MAGIC_SIZE, StreamReader, breakup_chunk, read_magic
2:   File "/Users/scpeters/ws/benchmark/src/benchmark/mcap/python/mcap/mcap/stream_reader.py", line 6, in <module>
2:     import lz4.frame  # type: ignore
2:     ^^^^^^^^^^^^^^^^
2: ModuleNotFoundError: No module named 'lz4'
2/4 Test #2: csv_BENCHMARK_boxes_dt ............***Failed    0.14 sec
~~~

File: boxes.cc [hunk 3]
```diff
@@ -40,17 +44,42 @@ void BoxesTest::Boxes(const std::string &_physicsEngine, double _dt,
...
   std::string world_path = boost::str(
       boost::format(
           "%1%/%2%/"
           "boxes_collision%3%_complex%4%_dt%5$.0e_modelCount%6%.world") %
       WORLDS_DIR_PATH % TEST_NAME % _collision % _complex % _dt % _modelCount);
 
+  std::string result_name = boost::str(
+    boost::format("%1%/%2%/MCAP/boxes_collision"
+            "%3%_complex%4%_dt%5$.0e_modelCount%6%_%7%.mcap") % RESULT_DIR_PATH %
+      TEST_NAME % _collision % _complex % _dt % _modelCount % _physicsEngine);
```
Comment (scpeters) [inline boxes.cc:58]: It's a little confusing to me to have the `mcap` initialization code mixed with the world initialization code. I would prefer to group the related code together, for example as follows:

<details>

patch:

~~~
diff --git a/boxes.cc b/boxes.cc
index 417e7a0..4927460 100644
--- a/boxes.cc
+++ b/boxes.cc
@@ -42,25 +42,13 @@ using namespace benchmark;
 void BoxesTest::Boxes(const std::string &_physicsEngine, double _dt,
                       int _modelCount, bool _collision, bool _complex) {
   ASSERT_GT(_modelCount, 0);
-  std::string world_erb_path =
-      boost::str(boost::format("%1%/boxes.world.erb") % WORLDS_DIR_PATH);
-  
-
-  std::string world_path = boost::str(
-      boost::format(
-          "%1%/%2%/"
-          "boxes_collision%3%_complex%4%_dt%5$.0e_modelCount%6%.world") %
-      WORLDS_DIR_PATH % TEST_NAME % _collision % _complex % _dt % _modelCount);
 
+  // Initialize MCAP logging
   std::string result_name = boost::str(
     boost::format("%1%/%2%/MCAP/boxes_collision"
             "%3%_complex%4%_dt%5$.0e_modelCount%6%_%7%.mcap") % RESULT_DIR_PATH %
       TEST_NAME % _collision % _complex % _dt % _modelCount % _physicsEngine);
 
-  std::string command = boost::str(
-      boost::format(
-          "erb collision=%1% complex=%2% dt=%3% modelCount=%4% %5% > %6%") %
-      _collision % _complex % _dt % _modelCount % world_erb_path % world_path);
   // for logging all the model in boxes_model_count test case 
   bool log_all = false;
   // mcap writer 
@@ -81,12 +69,27 @@ void BoxesTest::Boxes(const std::string &_physicsEngine, double _dt,
     channelId = channel.id;
   }
 
+  // Generate world file
+  std::string world_erb_path =
+      boost::str(boost::format("%1%/boxes.world.erb") % WORLDS_DIR_PATH);
+
+  std::string world_path = boost::str(
+      boost::format(
+          "%1%/%2%/"
+          "boxes_collision%3%_complex%4%_dt%5$.0e_modelCount%6%.world") %
+      WORLDS_DIR_PATH % TEST_NAME % _collision % _complex % _dt % _modelCount);
+
+  std::string command = boost::str(
+      boost::format(
+          "erb collision=%1% complex=%2% dt=%3% modelCount=%4% %5% > %6%") %
+      _collision % _complex % _dt % _modelCount % world_erb_path % world_path);
+
   // creating model with desired configuration
   auto model_check = system(command.c_str());
   // checking if model is created
   ASSERT_EQ(model_check, 0);
 
-  // Load a blank world (no ground plane)
+  // Load the generated world file
   Load(world_path, true, _physicsEngine);
   physics::WorldPtr world = physics::get_world("default");
   ASSERT_NE(world, nullptr);
~~~

what the reorganized code could look like:

~~~
void BoxesTest::Boxes(const std::string &_physicsEngine, double _dt,
                      int _modelCount, bool _collision, bool _complex) {
  ASSERT_GT(_modelCount, 0);

  // Initialize MCAP logging
  std::string result_name = boost::str(
    boost::format("%1%/%2%/MCAP/boxes_collision"
            "%3%_complex%4%_dt%5$.0e_modelCount%6%_%7%.mcap") % RESULT_DIR_PATH %
      TEST_NAME % _collision % _complex % _dt % _modelCount % _physicsEngine);

  // for logging all the model in boxes_model_count test case 
  bool log_all = false;
  // mcap writer 
  mcap::McapWriter writer;
  auto options = mcap::McapWriterOptions("");
  const auto s = writer.open(result_name.c_str(), options);
  ASSERT_EQ(s.ok(), true);

  mcap::ChannelId channelId;

  {
    mcap::Schema schema(
      "benchmark_proto.Boxes_msg", "protobuf",
      foxglove::BuildFileDescriptorSet(benchmark_proto::Boxes_msg::descriptor()).SerializeAsString());
    writer.addSchema(schema);
    mcap::Channel channel("boxes_states", "protobuf", schema.id);
    writer.addChannel(channel);
    channelId = channel.id;
  }

  // Generate world file
  std::string world_erb_path =
      boost::str(boost::format("%1%/boxes.world.erb") % WORLDS_DIR_PATH);

  std::string world_path = boost::str(
      boost::format(
          "%1%/%2%/"
          "boxes_collision%3%_complex%4%_dt%5$.0e_modelCount%6%.world") %
      WORLDS_DIR_PATH % TEST_NAME % _collision % _complex % _dt % _modelCount);

  std::string command = boost::str(
      boost::format(
          "erb collision=%1% complex=%2% dt=%3% modelCount=%4% %5% > %6%") %
      _collision % _complex % _dt % _modelCount % world_erb_path % world_path);

  // creating model with desired configuration
  auto model_check = system(command.c_str());
  // checking if model is created
  ASSERT_EQ(model_check, 0);

  // Load the generated world file
  Load(world_path, true, _physicsEngine);
  physics::WorldPtr world = physics::get_world("default");
  ASSERT_NE(world, nullptr);


~~~

</details>

Comment (yaswanth1701) [reply boxes.cc:58]: yeah, this makes more sense

**Comment by yaswanth1701** [comment]
> unfortunately the `css_*` tests are still failing for me, being unable to find the `lz4` python module
> 
> ```
> test 2
>     Start 2: csv_BENCHMARK_boxes_dt
> 
> 2: Test command: /usr/local/bin/python3 "/Users/scpeters/ws/benchmark/src/benchmark/tools/mcap_to_csv.py" "BENCHMARK_boxes_dt"
> 2: Working Directory: /Users/scpeters/ws/benchmark/src/benchmark/build
> 2: Test timeout computed to be: 10000000
> 2: Traceback (most recent call last):
> 2:   File "/Users/scpeters/ws/benchmark/src/benchmark/tools/mcap_to_csv.py", line 6, in <module>
> 2:     from mcap.reader import make_reader
> 2:   File "/Users/scpeters/ws/benchmark/src/benchmark/mcap/python/mcap/mcap/reader.py", line 37, in <module>
> 2:     from .stream_reader import MAGIC_SIZE, StreamReader, breakup_chunk, read_magic
> 2:   File "/Users/scpeters/ws/benchmark/src/benchmark/mcap/python/mcap/mcap/stream_reader.py", line 6, in <module>
> 2:     import lz4.frame  # type: ignore
> 2:     ^^^^^^^^^^^^^^^^
> 2: ModuleNotFoundError: No module named 'lz4'
> 2/4 Test #2: csv_BENCHMARK_boxes_dt ............***Failed    0.14 sec
> ```
can you try `pip install lz4`?
 because it's working on ubuntu (I tried with ubuntu docker image too)

**Comment by yaswanth1701** [comment]
Made a few changes to logging, should be much clear.

**Review by scpeters — COMMENTED** [review]
I needed to `pip install lz4 protobuf zstandard`; please add these instructions to the README

File: tools/mcap_to_csv.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                "quaternion_z",]
+
+CONFIGURATION  = ["physics_engine", "time_step", "complex", 
+                  "collisiion", "model_count","wall_time", 
+                  "log_multiple", "classname"]
+
+
+
+def get_file_names(result_folder):
+    result_dir = os.path.join("~", "benchmark","test_results", result_folder)
```
Comment (scpeters) [inline tools/mcap_to_csv.py:36]:
```suggestion
    result_dir = os.path.join(benchmark_dir,"test_results", result_folder)
```

[the remark this answers is not in the document]
Comment (yaswanth1701) [reply tools/mcap_to_csv.py]: yes, I made the changes to it.

**Review by scpeters — APPROVED** [review]

**Pull request merged**